controllers/osbuildenvconfig_controller.go

```diff
@@ -490,20 +660,323 @@ func (r *OSBuildEnvConfigReconciler) generateComposerService(serviceName, portNa
+	}
+
+	return nil, false, err
+}
+
+func (r *OSBuildEnvConfigReconciler) generateWorkerSSHKeysSecret(instance *osbuildv1alpha1.OSBuildEnvConfig, privateKeyBytes, publicKeyBytes []byte) (*corev1.Secret, error) {
+	immutable := true
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      workerSSHKeysSecretName,
+			Namespace: conf.GlobalConf.WorkingNamespace,
+		},
+		Immutable: &immutable,
+		StringData: map[string]string{
+			workerSSHKeysPrivateKeyKey: string(privateKeyBytes),
+			workerSSHKeysPublicKeyKey:  string(publicKeyBytes),
+		},
+	}
+
+	return secret, controllerutil.SetControllerReference(instance, secret, r.Scheme)
+}
+
+func (r *OSBuildEnvConfigReconciler) ensureWorkerVMExists(ctx context.Context, vmParameters *workerVMParameters, instance *osbuildv1alpha1.OSBuildEnvConfig) (bool, error) {
+	_, err := r.VirtualmachineRepository.Read(ctx, vmParameters.Name, conf.GlobalConf.WorkingNamespace)
+	if err == nil {
+		return false, nil
+	}
+
+	if errors.IsNotFound(err) {
+		workerVm, err := r.generateWorkerVN(vmParameters, instance)
+		if err != nil {
+			return false, err
+		}
+
+		err = r.VirtualmachineRepository.Create(ctx, workerVm)
+		if err != nil {
+			return false, err
+		}
+
+		return true, nil
+	}
+
+	return false, err
+}
+
+func (r *OSBuildEnvConfigReconciler) generateWorkerVN(vmParameters *workerVMParameters, instance *osbuildv1alpha1.OSBuildEnvConfig) (*kubevirtv1.VirtualMachine, error) {
+	buf, err := templates.LoadFromTemplateFile(workerVMTemplateFile, vmParameters)
+	if err != nil {
+		return nil, err
+	}
+
+	obj, groupVersionKind, err := serializer.NewCodecFactory(r.Scheme).UniversalDeserializer().Decode(buf.Bytes(), nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	if *groupVersionKind != kubevirtv1.VirtualMachineGroupVersionKind {
+		return nil, fmt.Errorf("failed to deserializer into a VirtualMachine CR")
+	}
+
+	vm, ok := obj.(*kubevirtv1.VirtualMachine)
+	if !ok {
+		return nil, fmt.Errorf("failed to cast into a VirtualMachine object")
+	}
+
+	return vm, controllerutil.SetControllerReference(instance, vm, r.Scheme)
+}
+
+func (r *OSBuildEnvConfigReconciler) ensureWorkerVMSSHServiceExists(ctx context.Context, workerName string, instance *osbuildv1alpha1.OSBuildEnvConfig) (bool, error) {
+	return r.ensureServiceExists(ctx, fmt.Sprintf(workerSSHServiceNameFormat, workerName), workerSSHPortName, 22, 22, map[string]string{"osbuild-worker": workerName}, instance)
+}
+
+func (r *OSBuildEnvConfigReconciler) ensureWorkerCertificateExists(ctx context.Context, reqLogger logr.Logger, instance *osbuildv1alpha1.OSBuildEnvConfig, workerName string) (bool, error) {
+	return r.ensureCertificateExists(ctx, reqLogger, instance, fmt.Sprintf(workerCertificateNameFormat, workerName), []string{workerName})
+}
+
+func (r *OSBuildEnvConfigReconciler) ensureWorkerConfigPlaybookExists(ctx context.Context, instance *osbuildv1alpha1.OSBuildEnvConfig) (bool, error) {
+	workerConfigPlaybookParams := struct {
```

Comment:
Maybe better to create a type for that struct? 
not sure, maybe for testing it will be easier?

Reply:
I don't really see the benefit, but I made the change